velox/expression/tests/ExpressionFuzzer.cpp

```diff
@@ -1045,20 +1062,108 @@ core::TypedExprPtr ExpressionFuzzer::generateExpression(
+        decimalNameStyle = 'b';
+        break;
+      }
+      if (variableName.find("i") != std::string::npos) {
+        decimalNameStyle = 'c';
+        break;
+      }
+    }
+  }
+
+  // To handle the constraints between input types and output types of a decimal
+  // function, extracts the input precisions and scales from decimal arguments,
+  // and bind them to integer variables.
+  std::unordered_map<std::string, int> decimalVariablesBindings;
+  column_index_t decimalColIndex = 1;
+  for (column_index_t i = 0; i < args.size(); ++i) {
+    const auto argType = args[i]->type();
+    if (argType->isDecimal()) {
+      const auto [p, s] = getDecimalPrecisionScale(*argType);
+      switch (decimalNameStyle) {
+        case 'a': {
+          decimalVariablesBindings["precision"] = p;
+          decimalVariablesBindings["scale"] = s;
+          break;
+        }
+        case 'b': {
+          const auto column = std::string(1, 'a' + i);
+          decimalVariablesBindings[column + "_precision"] = p;
+          decimalVariablesBindings[column + "_scale"] = s;
+          break;
+        }
+        case 'c': {
+          decimalVariablesBindings["i" + std::to_string(decimalColIndex)] = p;
+          decimalVariablesBindings
+              ["i" + std::to_string(decimalColIndex + kIntegerPairSize)] = s;
+          decimalColIndex++;
+          break;
+        }
+        default:
+          VELOX_UNSUPPORTED(
+              "Unsupported decimal name style {}.", decimalNameStyle);
+      }
+    }
+  }
+
+  // Calculates the matched return type through the argument types with argument
+  // type fuzzer, which evaluates the constraints internally.
+  if (integerConstrained && decimalVariablesBindings.size() > 0 && signature) {
+    ArgumentTypeFuzzer fuzzer{*signature, rng_, decimalVariablesBindings};
+    return fuzzer.fuzzReturnType();
+  }
+  return nullptr;
+}
+
 core::TypedExprPtr ExpressionFuzzer::getCallExprFromCallable(
     const CallableSignature& callable,
-    const TypePtr& type) {
+    const TypePtr& type,
+    const exec::FunctionSignature* signature) {
   auto args = getArgsForCallable(callable);
-  // Generate a CallTypedExpr with type because callable.returnType may not have
-  // the required field names.
-  return std::make_shared<core::CallTypedExpr>(type, args, callable.name);
+
+  // For a decimal function (especially a nested one), as argument precisions
+  // and scales are randomly generated, callable.returnType does not follow the
+  // required constraints, and the matched result type needs to be recalculated
+  // from the argument types. If function signature is provided, generates a
+  // constrained type to avoid breaking the constraints between input types and
+  // output types. Otherwise, generate a CallTypedExpr with type because
+  // callable.returnType may not have the required field names.
+  const auto constrainedType = getConstrainedOutputType(args, signature);
```

Comment:
nit: would it make sense to move any manipulation (enforcing constraints on the return type) to the caller function (`generateExpressionFromSignatureTemplate`) instead?